lib/kafka_ex/protocol/offset_commit.ex

```diff
@@ -29,6 +30,28 @@ defmodule KafkaEx.Protocol.OffsetCommit do
 
   @spec create_request(integer, binary, Request.t()) :: binary
   def create_request(correlation_id, client_id, offset_commit_request) do
+    create_request_by_version(correlation_id, client_id,
+      offset_commit_request, offset_commit_request.api_version)
+  end
+
+  defp create_request_by_version(
```

Comment:
Could we perhaps just collapse this down into a single function with an optional `api_version_number` parameter?

Reply:
The main reason is that I didn't want to change the signature of the `create_request` method. I.e. I wanted to keep the code changes to a minimum and not also make changes to server.ex and server_p0p8p2.ex.

Another thought was consistency amongst all `create_request` signatures but I just noticed that the create_request signatures do vary: [metadata has api_version](https://github.com/kafkaex/kafka_ex/blob/455fb652cfbf3554c885522a60329bd424c9aa60/lib/kafka_ex/protocol/metadata.ex#L148-L153), [fetch just takes fetch_request](https://github.com/kafkaex/kafka_ex/blob/455fb652cfbf3554c885522a60329bd424c9aa60/lib/kafka_ex/protocol/fetch.ex#L60) - taking correlation_id and client_id from the request, [heartbeat has correlation_id, client_id and request](https://github.com/kafkaex/kafka_ex/blob/455fb652cfbf3554c885522a60329bd424c9aa60/lib/kafka_ex/protocol/heartbeat.ex#L26) and [create_topics has api_version included](https://github.com/kafkaex/kafka_ex/blob/455fb652cfbf3554c885522a60329bd424c9aa60/lib/kafka_ex/protocol/create_topics.ex#L81-L86).

One minor point is that the way this is done (i.e. including api_version in the Request struct) would mean that the call becomes something like

    create_request(correlation_id, client_id, request, request.api_version)

so the api_version comes from the request, which is kinda doubling up when the request is already being passed in.

For me, it all starts with where to store the api_version. If it's stored directly in the struct defining the request, then passing it in explicitly, when the request is already being passed in, makes no sense. There are other ways to distinguish the various api_versions' from one another. For me, how the api_version is represented will define, in part, the signature for `create_request` across the board.

Sorry this got a bit ranty :)

Reply:
I think we can probably leave api_version out of the Request struct, since ideally we would actually check with the server to know what versions it supports before sending a request (for example, if the version isn't provided. We should also probably set a constraint on which versions we supports (as-is, if version 2 came out, it would send version 0 right now, which would probably not be desirable)

Reply:
Checking what the server/Kafka can and can't do doesn't cover all use-cases. I.e. if Kafka can store the consumer offsets in either Kafka or Zookeeper, then KafkaEx would always store in Kafka (since this represents the higher api version) even though the user might want to support legacy software that stores to Zookeeper.

That's why other Kafka libraries (i.e. python and nodejs) tend not to do so much magic and instead make the user decide which api versions they would like to use.

Something else that I noticed was that KafkaEx stores the offset of the last-read message not the offset of the next-to-read message. This then means that the consumer offsets are always one behind the offset of the topic, making it appear that the consumer still has a message to consumer. When requesting new messages, KafkaEx then adds one to the current consumer offset and all is well. Other libraries set the offset to the next-to-read offset and don't add one to the offset when requesting more messages. This then avoids the appearance that the consumer still has a message to consumer.

Reply:
When I read the python docs at https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html it says:

* *api_version (tuple) –*
Specify which Kafka API version to use. If set to None, the client will attempt to infer the broker version by probing various APIs. Different versions enable different functionality.
*Examples*
`(0, 9)` enables full group coordination features with automatic
partition assignment and rebalancing,
`(0, 8, 2)` enables kafka-storage offset commits with manual
partition assignment only,
`(0, 8, 1)` enables zookeeper-storage offset commits with manual
partition assignment only,
`(0, 8, 0)` enables basic functionality but requires manual
partition assignment and offset management.

We do want to provide a way to select where to store the offsets, since migration will be needed to get from zookeeper to kafka, but I think that could be a configuration setting that's more user friendly than having to know API version 0 vs 1 for a specific protocol message.

Reply:
Of course, having one central configuration to enable whatever use-case the user would needs is ideal. However you'll need to differentiate between protocol versions because requests and response vary between versions. 

What I meant when I mentioned python was something along the lines of the following:

```python
cli = KafkaClient(bootstrap_servers=broker_hosts, client_id='client_id')
request = ListGroupsRequest[1]()
for broker_id in broker_ids:
    future = cli.send(broker_id, request)
    cli.poll(future=future)
    ...
```

where the 1 in the line `ListGroupsRequest[1]()` defines which api version the user wants. If I wanted to use api version 0, then all I have to do is replace the 1 with a 0, or a 2 for that matter for version 2. The use-case I'm dealing with here is introspection of the Kafka and what offset values I have for various consumers.

If I was only doing consumption and production of messages, then I won't be interested in version numbers, I would only be interested in where the offset was stored and which partition has the data I need. And for that, one central configuration would be perfect.

At the moment, I'm only doing consumption with KafkaEx and for that, this change is sufficient since I only want to write my offsets to Kafka and not Zookeeper. If I wanted to do the same introspection that I'm doing with Python with KafkaEx, then that won't be as simple.

Reply:
So, we've had a long discussion here, but I think my feedback is that this function could be simplified by having a default parameter, without having to introduce a second function that takes a version number.
I also think we need to keep the version number out of the protocol struct, since we are already putting it as an argument in the function, and the other places that use a protocol api version do not add it to the protocol struct.